src/lib/consensus/proof_of_stake.ml

```diff
@@ -768,20 +816,44 @@ module Epoch_data = struct
     ; seed= Epoch_seed.initial
     ; start_checkpoint= Coda_base.State_hash.(of_hash zero)
     ; lock_checkpoint= Coda_base.State_hash.(of_hash zero)
-    ; length= Length.of_int 1 }
+    ; length= Length.of_int 1
+    ; participated_stake= Amount.zero }
 
   let update_pair (last_data, curr_data) epoch_length ~prev_epoch ~next_epoch
       ~prev_slot ~prev_protocol_state_hash ~proposer_vrf_result
-      ~snarked_ledger_hash ~total_currency =
+      ~snarked_ledger_hash ~total_currency ~delegator_addr ~local_state =
+    let open Or_error.Let_syntax in
     let open Epoch_ledger in
+    let delegator_account curr_epoch_data =
+      let epoch_snapshot =
+        if
+          Coda_base.Frozen_ledger_hash.equal curr_epoch_data.ledger.hash
+            genesis_ledger_hash
+        then Some local_state.Local_state.genesis_epoch_snapshot
+        else if
+          next_epoch > prev_epoch
+          || curr_epoch_data.length <= Length.of_int Constants.k
+        then local_state.curr_epoch_snapshot
+        else local_state.Local_state.last_epoch_snapshot
+      in
+      let%map epoch_ledger =
+        match epoch_snapshot with
+        | None ->
+            Or_error.error_string
+              "Got invalid snapshot when updating participated stake"
+        | Some snapshot -> Ok snapshot.ledger
+      in
```

Comment:
Can you assert that the epoch_ledger you selected is the one you expect from the consensus state?